File: Stack/Java/Minimum Add to Make Parentheses Valid.java
// Problem link: https://leetcode.com/problems/minimum-add-to-make-parentheses-valid/description
// Time: O(n)
// Space: O(1)
class Solution {
    public int minAddToMakeValid(String s) {
        int count = 0, result = 0;

        for(char chr : s.toCharArray()) {
            if(chr == '(') {
                count++;
            }
            else {
                if(count > 0) {
                    count--;
                }
                else {
                    result++;
                }
            }
        }

        return result + count;
    }
}

File: Stack/Python/Minimum Add to Make Parentheses Valid.py
# Problem link: https://leetcode.com/problems/minimum-add-to-make-parentheses-valid/description
# Time: O(n)
# Space: O(1)
class Solution:
    def minAddToMakeValid(self, s: str) -> int:
        result = count = 0

        for char in s:
            if char == '(':
                count += 1
            else:
                if count:
                    count -= 1
                else:
                    result += 1
        
        return result + count
